src/rgw/rgw_role.cc

```diff
@@ -0,0 +1,220 @@
+                  << id << ": " << cpp_strerror(-ret) << dendl;
+  }
+
+  // Delete name
+  ret = rgw_delete_system_obj(store, store->get_zone_params().roles_pool, name, NULL);
+  if (ret < 0) {
+    ldout(cct, 0) << "ERROR: deleting role name from pool: " << pool_name << ": "
+                  << name << ": " << cpp_strerror(-ret) << dendl;
+  }
+
+  return ret;
+}
+
+int RGWRole::list()
+{
+  string pool_name = store->get_zone_params().roles_pool.name;
+
+  int ret = read_name();
+  if (ret < 0) {
+    ldout(cct, 0) << "ERROR: finding role name in pool: " << pool_name << ": "
+                  << name << ": " << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  ret = read_info();
+  if (ret < 0) {
+    ldout(cct, 0) << "ERROR: finding role id in pool: " << pool_name << ": "
+                  << id << ": " << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  return 0;
+}
+
+void RGWRole::dump(Formatter *f) const
+{
+  encode_json("id", id , f);
+  encode_json("name", name , f);
+  encode_json("path", path, f);
+  encode_json("create_date", creation_date, f);
+}
+
+void RGWRole::decode_json(JSONObj *obj)
+{
+  JSONDecoder::decode_json("id", id, obj);
+  JSONDecoder::decode_json("name", name, obj);
+  JSONDecoder::decode_json("path", path, obj);
+  JSONDecoder::decode_json("create_date", creation_date, obj);
+}
+
+int RGWRole::read_id(const string& role_name, string& role_id)
+{
+  bufferlist bl;
+
+  RGWObjectCtx obj_ctx(store);
+  int ret = rgw_get_system_obj(store, obj_ctx, store->get_zone_params().roles_pool,
+                                  role_name, bl, NULL, NULL);
+  if (ret < 0) {
+    return ret;
+  }
+
+  string pool_name = store->get_zone_params().roles_pool.name;
+  RGWNameToId nameToId;
+  try {
+    bufferlist::iterator iter = bl.begin();
+    ::decode(nameToId, iter);
+  } catch (buffer::error& err) {
+    ldout(cct, 0) << "ERROR: failed to decode role from pool: " << pool_name << ": "
+                  << role_name << dendl;
+    return -EIO;
+  }
+  role_id = nameToId.obj_id;
+  return 0;
+}
+
+int RGWRole::read_info()
+{
+  string pool_name = store->get_zone_params().roles_pool.name;
+
+  bufferlist bl;
+  RGWObjectCtx obj_ctx(store);
+  int ret = rgw_get_system_obj(store, obj_ctx, store->get_zone_params().roles_pool,
+                                id, bl, NULL, NULL);
+  if (ret < 0) {
+    ldout(cct, 0) << "ERROR: failed reading role info from pool: " << pool_name <<
+                  ": " << id << ": " << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  try {
+    bufferlist::iterator iter = bl.begin();
+    ::decode(*this, iter);
+  } catch (buffer::error& err) {
+    ldout(cct, 0) << "ERROR: failed to decode role info from pool: " << pool_name <<
+                  ": " << id << dendl;
+    return -EIO;
+  }
+
+  return 0;
+}
+
+int RGWRole::read_name()
+{
+  string pool_name = store->get_zone_params().roles_pool.name;
```

Comment:
shouldn't need to make a copy of the string here. consider changing this to `auto& pool = store->get_zone_params().roles_pool`, so it be used for both the `rgw_get_system_obj()` call, and to access the name as `pool.name`. can do similar in the other functions too